src/Raven.Server/Documents/Patch/V8/ScriptRunner.cs

```diff
@@ -0,0 +1,1700 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using V8.Net;
+using Raven.Server.Extensions.V8;
+using Raven.Client;
+using Raven.Client.Documents.Indexes.Spatial;
+using Raven.Client.Documents.Session.TimeSeries;
+using Raven.Client.Exceptions.Documents;
+using Raven.Client.Exceptions.Documents.Patching;
+using Raven.Server.ServerWide;
+using Raven.Server.Documents.Indexes.Static.Spatial;
+using Raven.Server.Documents.Queries.Results.TimeSeries;
+using Raven.Server.Documents.Queries.Timings;
+using Raven.Server.Documents.TimeSeries;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow;
+using Sparrow.Extensions;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Spatial4n.Core.Distance;
+using ExpressionType = System.Linq.Expressions.ExpressionType;
+using Raven.Server.Documents.Patch.V8;
+
+namespace Raven.Server.Documents.Patch
+{
+    public partial class ScriptRunner
+    {
+        public static unsafe DateTime GetDateArg(InternalHandle arg, string signature, string argName)
+        {
+            if (arg.IsDate)
+                return arg.AsDate;
+
+            if (arg.IsStringEx == false)
+                ThrowInvalidDateArgument();
+
+            var s = arg.AsString;
+            fixed (char* pValue = s)
+            {
+                var result = LazyStringParser.TryParseDateTime(pValue, s.Length, out DateTime dt, out _);
+                if (result != LazyStringParser.Result.DateTime)
+                    ThrowInvalidDateArgument();
+
+                return dt;
+            }
+
+            void ThrowInvalidDateArgument() =>
+                throw new ArgumentException($"{signature} : {argName} must be of type 'DateInstance' or a DateTime string. {GetTypes(arg)}");
+        }
+
+        private static DateTime GetTimeSeriesDateArg(InternalHandle arg, string signature, string argName)
+        {
+            if (arg.IsDate)
+                return arg.AsDate;
+
+            if (arg.IsStringEx == false)
+                throw new ArgumentException($"{signature} : {argName} must be of type 'DateInstance' or a DateTime string. {GetTypes(arg)}");
+
+            return TimeSeriesRetriever.ParseDateTime(arg.AsString);
+        }
+        
+        private static string GetTypes(InternalHandle value) => $"JintType({value.ValueType}) .NETType({value.GetType().Name})";
+        
+        public partial class SingleRun
+        {
+            public V8EngineEx ScriptEngineExV8;
+            public V8Engine ScriptEngineV8;
+            public JavaScriptUtilsV8 JavaScriptUtilsV8;
+
+            public void InitializeV8()
+            {
+                ScriptEngineExV8 = new V8EngineEx(_jsOptions);
+                ScriptEngineV8 = ScriptEngineExV8;
+                ScriptEngine = ScriptEngineExV8;
+
```

Comment:
I don't understand why have this three times

Reply:
removed scriptEngineExV8 as it wasn't used outside this method
renamed ScriptEngine to ScriptEngineHandle

now we have:
IJavaScriptEngineHandle ScriptEngineHandle - generic level (abstract Jint and V8)
ScriptEngineV8 - original V8Engine level